File: report.py
"""
Report Generation Module
Handles receipt generation and sales reports
"""

import os
import json
from datetime import datetime

class ReportGenerator:
    def __init__(self, receipts_file="data/receipts.txt", receipt_details_file="data/receipt_details.txt"):
        self.receipts_file = receipts_file
        self.receipt_details_file = receipt_details_file
        self.receipts = []
        self.receipt_details = []
        self.load_receipts()
        
    def load_receipts(self):
        """Load receipts from file"""
        if not os.path.exists(os.path.dirname(self.receipts_file)):
            os.makedirs(os.path.dirname(self.receipts_file))
            
        # Load receipt headers
        if os.path.exists(self.receipts_file):
            try:
                with open(self.receipts_file, 'r') as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith('#'):
                            parts = line.split(',')
                            if len(parts) >= 4:
                                receipt = {
                                    'receipt_id': parts[0].strip(),
                                    'customer_id': parts[1].strip(),
                                    'date': parts[2].strip(),
                                    'total_amount': float(parts[3].strip())
                                }
                                self.receipts.append(receipt)
            except Exception as e:
                print(f"Error loading receipts: {e}")
        else:
            # Create empty receipts file
            with open(self.receipts_file, 'w') as f:
                f.write("# ReceiptID, CustID, Date, TotalAmount\n")
                
        # Load receipt details
        if os.path.exists(self.receipt_details_file):
            try:
                with open(self.receipt_details_file, 'r') as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith('#'):
                            parts = line.split(',')
                            if len(parts) >= 5:
                                detail = {
                                    'receipt_id': parts[0].strip(),
                                    'item_id': parts[1].strip(),
                                    'item_name': parts[2].strip(),
                                    'quantity': int(parts[3].strip()),
                                    'price': float(parts[4].strip()),
                                    'total': float(parts[5].strip()) if len(parts) > 5 else float(parts[4].strip()) * int(parts[3].strip())
                                }
                                self.receipt_details.append(detail)
            except Exception as e:
                print(f"Error loading receipt details: {e}")
        else:
            # Create empty receipt details file
            with open(self.receipt_details_file, 'w') as f:
                f.write("# ReceiptID, ItemID, ItemName, Quantity, Price, Total\n")
                
    def generate_receipt(self, customer, cart_items):
        """Generate receipt for customer purchase"""
        if not cart_items:
            return None
            
        # Generate receipt ID
        receipt_id = self.get_next_receipt_id()
        current_time = datetime.now()
        date_str = current_time.strftime("%Y-%m-%d")
        time_str = current_time.strftime("%H:%M:%S")
        
        # Calculate total
        total_amount = 0
        receipt_details = []
        
        for cart_item in cart_items:
            item = cart_item['item']
            quantity = cart_item['quantity']
            line_total = item['price'] * quantity
            total_amount += line_total
            
            detail = {
                'receipt_id': receipt_id,
                'item_id': item['id'],
                'item_name': item['name'],
                'quantity': quantity,
                'price': item['price'],
                'total': line_total
            }
            receipt_details.append(detail)
            
        # Create receipt record
        receipt = {
            'receipt_id': receipt_id,
            'customer_id': customer['id'],
            'date': date_str,
            'total_amount': total_amount
        }
        
        # Save receipt
        self.receipts.append(receipt)
        self.receipt_details.extend(receipt_details)
        self.save_receipts()
        
        # Display receipt
        self.display_receipt(customer, receipt_details, total_amount, current_time)
        
        return receipt_id
        
    def display_receipt(self, customer, receipt_details, total_amount, timestamp):
        """Display formatted receipt"""
        print("\n" + "="*60)
        print("                    XYZ RETAIL STORE")
        print("                   CUSTOMER RECEIPT")
        print("="*60)
        print(f"Date: {timestamp.strftime('%Y-%m-%d')}          Time: {timestamp.strftime('%H:%M:%S')}")
        print(f"Receipt ID: {receipt_details[0]['receipt_id']}")
        print("-"*60)
        print(f"Customer: {customer['name']}")
        print(f"Phone: {customer['phone']}")
        print(f"Email: {customer['email']}")
        print("-"*60)
        print(f"{'Item ID':<8} {'Item Name':<20} {'Qty':<5} {'Price':<10} {'Total':<10}")
        print("-"*60)
        
        for detail in receipt_details:
            print(f"{detail['item_id']:<8} {detail['item_name']:<20} {detail['quantity']:<5} ${detail['price']:<9.2f} ${detail['total']:<9.2f}")
            
        print("-"*60)
        print(f"{'TOTAL AMOUNT':<50} ${total_amount:.2f}")
        print("="*60)
        print("            Thank you for shopping with us!")
        print("="*60)
        
    def get_next_receipt_id(self):
        """Generate next receipt ID"""
        if not self.receipts:
            return "R001"
            
        max_id = 0
        for receipt in self.receipts:
            try:
                receipt_num = int(receipt['receipt_id'][1:])  # Remove 'R' prefix
                if receipt_num > max_id:
                    max_id = receipt_num
            except ValueError:
                continue
                
        return f"R{max_id + 1:03d}"
        
    def save_receipts(self):
        """Save receipts to files"""
        try:
            # Save receipt headers
            with open(self.receipts_file, 'w') as f:
                f.write("# ReceiptID, CustID, Date, TotalAmount\n")
                for receipt in self.receipts:
                    f.write(f"{receipt['receipt_id']}, {receipt['customer_id']}, {receipt['date']}, {receipt['total_amount']}\n")
                    
            # Save receipt details
            with open(self.receipt_details_file, 'w') as f:
                f.write("# ReceiptID, ItemID, ItemName, Quantity, Price, Total\n")
                for detail in self.receipt_details:
                    f.write(f"{detail['receipt_id']}, {detail['item_id']}, {detail['item_name']}, {detail['quantity']}, {detail['price']}, {detail['total']}\n")
                    
        except Exception as e:
            print(f"Error saving receipts: {e}")
            
    def daily_sales_report(self, date):
        """Generate daily sales report"""
        print(f"\n{'='*50}")
        print(f"           DAILY SALES REPORT - {date}")
        print(f"{'='*50}")
        
        daily_receipts = [r for r in self.receipts if r['date'] == date]
        
        if not daily_receipts:
            print("No sales found for this date.")
            return
            
        total_sales = 0
        total_transactions = len(daily_receipts)
        
        print(f"{'Receipt ID':<12} {'Customer ID':<12} {'Amount':<12}")
        print("-"*40)
        
        for receipt in daily_receipts:
            total_sales += receipt['total_amount']
            print(f"{receipt['receipt_id']:<12} {receipt['customer_id']:<12} ${receipt['total_amount']:<11.2f}")
            
        print("-"*40)
        print(f"Total Transactions: {total_transactions}")
        print(f"Total Sales: ${total_sales:.2f}")
        print(f"Average Transaction: ${total_sales/total_transactions:.2f}" if total_transactions > 0 else "Average Transaction: $0.00")
        print("="*50)
        
    def customer_purchase_history(self, customer_id):
        """Generate customer purchase history"""
        print(f"\n{'='*60}")
        print(f"        CUSTOMER PURCHASE HISTORY - ID: {customer_id}")
        print(f"{'='*60}")
        
        customer_receipts = [r for r in self.receipts if r['customer_id'] == customer_id]
        
        if not customer_receipts:
            print("No purchase history found for this customer.")
            return
            
        total_spent = 0
        
        for receipt in customer_receipts:
            total_spent += receipt['total_amount']
            print(f"\nReceipt ID: {receipt['receipt_id']} | Date: {receipt['date']} | Amount: ${receipt['total_amount']:.2f}")
            
            # Show items for this receipt
            receipt_items = [d for d in self.receipt_details if d['receipt_id'] == receipt['receipt_id']]
            for item in receipt_items:
                print(f"  - {item['item_name']} x{item['quantity']} @ ${item['price']:.2f} = ${item['total']:.2f}")
                
        print("-"*60)
        print(f"Total Purchases: {len(customer_receipts)}")
        print(f"Total Amount Spent: ${total_spent:.2f}")
        print("="*60)
        
    def item_sales_summary(self):
        """Generate item sales summary"""
        print(f"\n{'='*60}")
        print(f"                ITEM SALES SUMMARY")
        print(f"{'='*60}")
        
        if not self.receipt_details:
            print("No sales data available.")
            return
            
        # Aggregate item sales
        item_stats = {}
        for detail in self.receipt_details:
            item_id = detail['item_id']
            if item_id not in item_stats:
                item_stats[item_id] = {
                    'name': detail['item_name'],
                    'quantity_sold': 0,
                    'total_revenue': 0
                }
            item_stats[item_id]['quantity_sold'] += detail['quantity']
            item_stats[item_id]['total_revenue'] += detail['total']
            
        # Sort by revenue
        sorted_items = sorted(item_stats.items(), key=lambda x: x[1]['total_revenue'], reverse=True)
        
        print(f"{'Item ID':<8} {'Item Name':<20} {'Qty Sold':<10} {'Revenue':<12}")
        print("-"*60)
        
        total_revenue = 0
        total_quantity = 0
        
        for item_id, stats in sorted_items:
            total_revenue += stats['total_revenue']
            total_quantity += stats['quantity_sold']
            print(f"{item_id:<8} {stats['name']:<20} {stats['quantity_sold']:<10} ${stats['total_revenue']:<11.2f}")
            
        print("-"*60)
        print(f"Total Items Sold: {total_quantity}")
        print(f"Total Revenue: ${total_revenue:.2f}")
        print("="*60)

File: .ccls-cache/@home@runner@workspace/report.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "report.h"
#include "item.h"

// Global variables
Receipt receipts[MAX_RECEIPTS];
ReceiptDetail receipt_details[MAX_RECEIPT_DETAILS];
int receipt_count = 0;
int receipt_detail_count = 0;

void load_receipts() {
    // Load receipt headers
    FILE *file = fopen("data/receipts.dat", "r");
    if (file != NULL) {
        receipt_count = 0;
        while (fscanf(file, "%9[^,],%d,%14[^,],%f\n", 
                      receipts[receipt_count].receipt_id, 
                      &receipts[receipt_count].customer_id, 
                      receipts[receipt_count].date, 
                      &receipts[receipt_count].total_amount) == 4) {
            receipt_count++;
            if (receipt_count >= MAX_RECEIPTS) break;
        }
        fclose(file);
    }
    
    // Load receipt details
    file = fopen("data/receipt_details.dat", "r");
    if (file != NULL) {
        receipt_detail_count = 0;
        while (fscanf(file, "%9[^,],%d,%49[^,],%d,%f,%f\n", 
                      receipt_details[receipt_detail_count].receipt_id, 
                      &receipt_details[receipt_detail_count].item_id, 
                      receipt_details[receipt_detail_count].item_name, 
                      &receipt_details[receipt_detail_count].quantity, 
                      &receipt_details[receipt_detail_count].price, 
                      &receipt_details[receipt_detail_count].total) == 6) {
            receipt_detail_count++;
            if (receipt_detail_count >= MAX_RECEIPT_DETAILS) break;
        }
        fclose(file);
    }
}

void save_receipts() {
    // Save receipt headers
    FILE *file = fopen("data/receipts.dat", "w");
    if (file != NULL) {
        for (int i = 0; i < receipt_count; i++) {
            fprintf(file, "%s,%d,%s,%.2f\n", 
                    receipts[i].receipt_id, 
                    receipts[i].customer_id, 
                    receipts[i].date, 
                    receipts[i].total_amount);
        }
        fclose(file);
    }
    
    // Save receipt details
    file = fopen("data/receipt_details.dat", "w");
    if (file != NULL) {
        for (int i = 0; i < receipt_detail_count; i++) {
            fprintf(file, "%s,%d,%s,%d,%.2f,%.2f\n", 
                    receipt_details[i].receipt_id, 
                    receipt_details[i].item_id, 
                    receipt_details[i].item_name, 
                    receipt_details[i].quantity, 
                    receipt_details[i].price, 
                    receipt_details[i].total);
        }
        fclose(file);
    }
}

char* generate_receipt(Customer* customer) {
    if (is_cart_empty()) {
        printf("Cart is empty! Cannot generate receipt.\n");
        return NULL;
    }
    
    static char receipt_id[10];
    strcpy(receipt_id, get_next_receipt_id());
    
    char date_str[MAX_DATE_LEN];
    get_current_date(date_str);
    
    float total_amount = get_cart_total();
    
    // Create receipt record
    Receipt new_receipt;
    strcpy(new_receipt.receipt_id, receipt_id);
    new_receipt.customer_id = customer->id;
    strcpy(new_receipt.date, date_str);
    new_receipt.total_amount = total_amount;
    
    receipts[receipt_count] = new_receipt;
    receipt_count++;
    
    // Create receipt details
    for (int i = 0; i < cart.count; i++) {
        ReceiptDetail detail;
        strcpy(detail.receipt_id, receipt_id);
        detail.item_id = cart.items[i].item.id;
        strcpy(detail.item_name, cart.items[i].item.name);
        detail.quantity = cart.items[i].quantity;
        detail.price = cart.items[i].item.price;
        detail.total = cart.items[i].item.price * cart.items[i].quantity;
        
        receipt_details[receipt_detail_count] = detail;
        receipt_detail_count++;
        
        // Update stock
        update_item_stock(cart.items[i].item.id, cart.items[i].quantity);
    }
    
    save_receipts();
    display_receipt(customer, receipt_id, total_amount);
    
    return receipt_id;
}

void display_receipt(Customer* customer, const char* receipt_id, float total) {
    time_t t = time(NULL);
    struct tm tm = *localtime(&t);
    
    printf("\n==================================================\n");
    printf("                XYZ RETAIL STORE\n");
    printf("               CUSTOMER RECEIPT\n");
    printf("==================================================\n");
    printf("Date: %04d-%02d-%02d        Time: %02d:%02d:%02d\n", 
           tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
           tm.tm_hour, tm.tm_min, tm.tm_sec);
    printf("Receipt ID: %s\n", receipt_id);
    printf("--------------------------------------------------\n");
    printf("Customer: %s\n", customer->name);
    printf("Phone: %s\n", customer->phone);
    printf("Email: %s\n", customer->email);
    printf("--------------------------------------------------\n");
    printf("%-8s %-20s %-5s %-10s %-10s\n", "Item ID", "Item Name", "Qty", "Price", "Total");
    printf("--------------------------------------------------\n");
    
    for (int i = 0; i < cart.count; i++) {
        float line_total = cart.items[i].item.price * cart.items[i].quantity;
        printf("%-8d %-20s %-5d $%-9.2f $%-9.2f\n", 
               cart.items[i].item.id, 
               cart.items[i].item.name, 
               cart.items[i].quantity, 
               cart.items[i].item.price, 
               line_total);
    }
    
    printf("--------------------------------------------------\n");
    printf("%-40s $%.2f\n", "TOTAL AMOUNT", total);
    printf("==================================================\n");
    printf("           Thank you for shopping with us!\n");
    printf("==================================================\n");
}

char* get_next_receipt_id() {
    static char next_id[10];
    int max_num = 0;
    
    for (int i = 0; i < receipt_count; i++) {
        int num;
        if (sscanf(receipts[i].receipt_id, "R%d", &num) == 1) {
            if (num > max_num) {
                max_num = num;
            }
        }
    }
    
    sprintf(next_id, "R%03d", max_num + 1);
    return next_id;
}

void get_current_date(char* date_str) {
    time_t t = time(NULL);
    struct tm tm = *localtime(&t);
    sprintf(date_str, "%04d-%02d-%02d", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday);
}

void get_current_time(char* time_str) {
    time_t t = time(NULL);
    struct tm tm = *localtime(&t);
    sprintf(time_str, "%02d:%02d:%02d", tm.tm_hour, tm.tm_min, tm.tm_sec);
}

void daily_sales_report(const char* date) {
    printf("\n==================================================\n");
    printf("           DAILY SALES REPORT - %s\n", date);
    printf("==================================================\n");
    
    float total_sales = 0;
    int total_transactions = 0;
    
    printf("%-12s %-12s %-12s\n", "Receipt ID", "Customer ID", "Amount");
    printf("------------------------------------------\n");
    
    for (int i = 0; i < receipt_count; i++) {
        if (strcmp(receipts[i].date, date) == 0) {
            total_sales += receipts[i].total_amount;
            total_transactions++;
            printf("%-12s %-12d $%-11.2f\n", 
                   receipts[i].receipt_id, 
                   receipts[i].customer_id, 
                   receipts[i].total_amount);
        }
    }
    
    if (total_transactions == 0) {
        printf("No sales found for this date.\n");
    } else {
        printf("------------------------------------------\n");
        printf("Total Transactions: %d\n", total_transactions);
        printf("Total Sales: $%.2f\n", total_sales);
        printf("Average Transaction: $%.2f\n", total_sales / total_transactions);
    }
    printf("==================================================\n");
}

void customer_purchase_history(int customer_id) {
    printf("\n==================================================\n");
    printf("    CUSTOMER PURCHASE HISTORY - ID: %d\n", customer_id);
    printf("==================================================\n");
    
    float total_spent = 0;
    int purchase_count = 0;
    
    for (int i = 0; i < receipt_count; i++) {
        if (receipts[i].customer_id == customer_id) {
            total_spent += receipts[i].total_amount;
            purchase_count++;
            
            printf("\nReceipt ID: %s | Date: %s | Amount: $%.2f\n", 
                   receipts[i].receipt_id, 
                   receipts[i].date, 
                   receipts[i].total_amount);
            
            // Show items for this receipt
            for (int j = 0; j < receipt_detail_count; j++) {
                if (strcmp(receipt_details[j].receipt_id, receipts[i].receipt_id) == 0) {
                    printf("  - %s x%d @ $%.2f = $%.2f\n", 
                           receipt_details[j].item_name, 
                           receipt_details[j].quantity, 
                           receipt_details[j].price, 
                           receipt_details[j].total);
                }
            }
        }
    }
    
    if (purchase_count == 0) {
        printf("No purchase history found for this customer.\n");
    } else {
        printf("--------------------------------------------------\n");
        printf("Total Purchases: %d\n", purchase_count);
        printf("Total Amount Spent: $%.2f\n", total_spent);
    }
    printf("==================================================\n");
}

void item_sales_summary() {
    printf("\n==================================================\n");
    printf("                ITEM SALES SUMMARY\n");
    printf("==================================================\n");
    
    if (receipt_detail_count == 0) {
        printf("No sales data available.\n");
        return;
    }
    
    typedef struct {
        int item_id;
        char item_name[MAX_NAME_LEN];
        int quantity_sold;
        float total_revenue;
    } ItemStats;
    
    ItemStats stats[MAX_ITEMS];
    int stat_count = 0;
    
    // Aggregate item sales
    for (int i = 0; i < receipt_detail_count; i++) {
        int found = 0;
        for (int j = 0; j < stat_count; j++) {
            if (stats[j].item_id == receipt_details[i].item_id) {
                stats[j].quantity_sold += receipt_details[i].quantity;
                stats[j].total_revenue += receipt_details[i].total;
                found = 1;
                break;
            }
        }
        
        if (!found) {
            stats[stat_count].item_id = receipt_details[i].item_id;
            strcpy(stats[stat_count].item_name, receipt_details[i].item_name);
            stats[stat_count].quantity_sold = receipt_details[i].quantity;
            stats[stat_count].total_revenue = receipt_details[i].total;
            stat_count++;
        }
    }
    
    // Simple bubble sort by revenue (descending)
    for (int i = 0; i < stat_count - 1; i++) {
        for (int j = 0; j < stat_count - i - 1; j++) {
            if (stats[j].total_revenue < stats[j + 1].total_revenue) {
                ItemStats temp = stats[j];
                stats[j] = stats[j + 1];
                stats[j + 1] = temp;
            }
        }
    }
    
    printf("%-8s %-20s %-10s %-12s\n", "Item ID", "Item Name", "Qty Sold", "Revenue");
    printf("--------------------------------------------------\n");
    
    float total_revenue = 0;
    int total_quantity = 0;
    
    for (int i = 0; i < stat_count; i++) {
        total_revenue += stats[i].total_revenue;
        total_quantity += stats[i].quantity_sold;
        printf("%-8d %-20s %-10d $%-11.2f\n", 
               stats[i].item_id, 
               stats[i].item_name, 
               stats[i].quantity_sold, 
               stats[i].total_revenue);
    }
    
    printf("--------------------------------------------------\n");
    printf("Total Items Sold: %d\n", total_quantity);
    printf("Total Revenue: $%.2f\n", total_revenue);
    printf("==================================================\n");
}